yarpl/include/yarpl/flowable/PublishProcessor.h

```diff
@@ -0,0 +1,234 @@
+    DCHECK(publishers != kErroredPublishers);
+
+    for (const auto& publisher : *publishers) {
+      publisher->onError(ex);
+    }
+  }
+
+  void onComplete() override {
+    auto publishers = kCompletedPublishers;
+    publishers_.swap(publishers);
+    DCHECK(publishers != kCompletedPublishers);
+    DCHECK(publishers != kErroredPublishers);
+
+    for (const auto& publisher : *publishers) {
+      publisher->onComplete();
+    }
+  }
+
+ private:
+  std::shared_ptr<const PublishersVector> tryAddPublisher(
+      std::shared_ptr<PublisherSubscription> subscriber) {
+    while (true) {
+      auto oldPublishers = publishers_.copy();
+      if (oldPublishers == kCompletedPublishers ||
+          oldPublishers == kErroredPublishers) {
+        return oldPublishers;
+      }
+
+      auto newPublishers = std::make_shared<PublishersVector>();
+      newPublishers->reserve(oldPublishers->size() + 1);
+      newPublishers->insert(
+          newPublishers->begin(),
+          oldPublishers->cbegin(),
+          oldPublishers->cend());
+      newPublishers->push_back(subscriber);
+
+      auto locked = publishers_.lock();
+      if (*locked == oldPublishers) {
+        *locked = newPublishers;
+        return newPublishers;
+      }
+      // else the vector changed so we will have to do it again
+    }
+  }
+
+  void removePublisher(PublisherSubscription* subscriber) {
+    while (true) {
+      auto oldPublishers = publishers_.copy();
+
+      auto removingItem = std::find_if(
+          oldPublishers->cbegin(),
+          oldPublishers->cend(),
+          [&](const auto& publisherPtr) {
+            return publisherPtr.get() == subscriber;
+          });
+
+      if (removingItem == oldPublishers->cend()) {
+        // not found anymore
+        return;
+      }
+
+      auto newPublishers = std::make_shared<PublishersVector>();
+      newPublishers->reserve(oldPublishers->size() - 1);
+      newPublishers->insert(
+          newPublishers->begin(), oldPublishers->cbegin(), removingItem);
+      newPublishers->insert(
+          newPublishers->end(), std::next(removingItem), oldPublishers->cend());
+
+      auto locked = publishers_.lock();
+      if (*locked == oldPublishers) {
+        *locked = std::move(newPublishers);
+        return;
+      }
+      // else the vector changed so we will have to do it again
+    }
+  }
+
+  class PublisherSubscription : public Subscription {
+   public:
+    PublisherSubscription(
+        std::shared_ptr<Subscriber<T>> subscriber,
+        PublishProcessor* processor)
+        : subscriber_(std::move(subscriber)), processor_(processor) {}
+
+    // cancel may race with terminate(), but the
+    // PublishProcessor::removePublisher will take care of that the race with
+    // on{Next, Error, Complete} methods is allowed by the spec
+    void cancel() override {
+      credits_ = credits::kCanceled;
+      processor_->removePublisher(this);
+    }
+
+    // we don't care about races with this method
```

Comment:
Hm, why not?

Reply:
IMO we shouldn't be allowing requesting and signaling methods to race at all. We end up with implementations that need to pessimistically lock (which introduces both runtime overhead, and implementation complexity considerably). I think it's possible to have an architecture which is explicit in scheduling work on different executors and only performs synchronization at those boundaries (eg, `subscribe/observeOn`).